File: internal/contextpack/diff.go
package contextpack

import (
	"strconv"
	"strings"
)

// diffLines is what a unified diff says per path: the head-side lines it
// added, the head-side lines it shows at all (added plus context), and the
// base-side lines it removed.
type diffLines struct {
	added   map[string][]int
	shown   map[string]map[int]bool
	removed map[string][]int
}

// parseDiff walks a unified diff once. Paths are head-side for added and
// shown, base-side for removed; a rename therefore keys the two sides
// differently, which is what the two trees need.
func parseDiff(diff string) diffLines {
	d := diffLines{added: map[string][]int{}, shown: map[string]map[int]bool{}, removed: map[string][]int{}}
	var oldPath, newPath string
	var oldLine, newLine int
	inHunk := false
	for l := range strings.SplitSeq(diff, "\n") {
		switch {
		case strings.HasPrefix(l, "--- "):
			oldPath = stripPrefix(l[4:], "a/")
			inHunk = false
		case strings.HasPrefix(l, "+++ "):
			newPath = stripPrefix(l[4:], "b/")
			inHunk = false
		case strings.HasPrefix(l, "@@"):
			oldLine, newLine = hunkStarts(l)
			inHunk = oldLine > 0 && newLine > 0
		case !inHunk:
		case strings.HasPrefix(l, "+"):
			if newPath != "" {
				d.added[newPath] = append(d.added[newPath], newLine)
				d.show(newPath, newLine)
			}
			newLine++
		case strings.HasPrefix(l, "-"):
			if oldPath != "" {
				d.removed[oldPath] = append(d.removed[oldPath], oldLine)
			}
			oldLine++
		case strings.HasPrefix(l, " "):
			if newPath != "" {
				d.show(newPath, newLine)
			}
			oldLine++
			newLine++
		case strings.HasPrefix(l, "\\"):
		default:
			inHunk = false
		}
	}
	return d
}

func (d *diffLines) show(path string, line int) {
	if d.shown[path] == nil {
		d.shown[path] = map[int]bool{}
	}
	d.shown[path][line] = true
}

func stripPrefix(p, prefix string) string {
	p = strings.TrimSpace(p)
	if p == "/dev/null" {
		return ""
	}
	return strings.TrimPrefix(p, prefix)
}

// hunkStarts reads "@@ -a,b +c,d @@" and returns a and c.
func hunkStarts(header string) (oldStart, newStart int) {
	fields := strings.Fields(header)
	if len(fields) < 3 {
		return 0, 0
	}
	return startOf(fields[1]), startOf(fields[2])
}

func startOf(field string) int {
	field = strings.TrimLeft(field, "-+")
	if i := strings.IndexByte(field, ','); i >= 0 {
		field = field[:i]
	}
	n, err := strconv.Atoi(field)
	if err != nil {
		return 0
	}
	return n
}

// runs groups sorted line numbers into ranges, merging neighbours closer
// than gap lines.
func runs(lines []int, gap int) [][2]int {
	var out [][2]int
	for _, l := range lines {
		if n := len(out); n > 0 && l-out[n-1][1] <= gap {
			out[n-1][1] = l
			continue
		}
		out = append(out, [2]int{l, l})
	}
	return out
}

// Hunks returns the head-side text of each hunk in a unified diff: added
// and context lines, without the removed ones, so each string reads like
// the region as it now is. Hunks are keyed by their head path.
func Hunks(diff string) []Hunk {
	var out []Hunk
	var path string
	var cur *Hunk
	flush := func() {
		if cur != nil && strings.TrimSpace(cur.Text) != "" {
			out = append(out, *cur)
		}
		cur = nil
	}
	for l := range strings.SplitSeq(diff, "\n") {
		switch {
		case strings.HasPrefix(l, "+++ "):
			flush()
			path = stripPrefix(l[4:], "b/")
		case strings.HasPrefix(l, "--- "), strings.HasPrefix(l, "diff --git "):
			flush()
		case strings.HasPrefix(l, "@@"):
			flush()
			if path != "" {
				cur = &Hunk{Path: path}
			}
		case cur == nil:
		case strings.HasPrefix(l, "+"), strings.HasPrefix(l, " "):
			cur.Text += l[1:] + "\n"
		}
	}
	flush()
	return out
}

// Hunk is one hunk's head-side text.
type Hunk struct {
	Path string
	Text string
}

File: internal/web/src/lib/diff.ts
// Parses a unified `git diff` into files and numbered lines so the diff tab
// can anchor findings under the new-side line they point at.

export type LineKind = 'add' | 'del' | 'ctx' | 'hunk' | 'meta';

export interface DiffLine {
  kind: LineKind;
  text: string;
  oldNo: number | null;
  newNo: number | null;
}

export interface DiffFile {
  oldPath: string;
  newPath: string;
  // path is the side a reader identifies the file by: the new path, or the
  // old one for a deletion.
  path: string;
  header: string[];
  lines: DiffLine[];
  added: number;
  removed: number;
}

const HUNK = /^@@ -(\d+)(?:,\d+)? \+(\d+)(?:,\d+)? @@/;

function stripPrefix(p: string): string {
  if (p === '/dev/null') return '';
  return p.replace(/^[ab]\//, '');
}

export function parseDiff(src: string): DiffFile[] {
  const files: DiffFile[] = [];
  let cur: DiffFile | undefined;
  let oldNo = 0;
  let newNo = 0;
  let inHunk = false;
  for (const text of src.split('\n')) {
    const git = /^diff --git a\/(.*) b\/(.*)$/.exec(text);
    if (git) {
      cur = { oldPath: git[1] ?? '', newPath: git[2] ?? '', path: git[2] ?? '', header: [text], lines: [], added: 0, removed: 0 };
      files.push(cur);
      inHunk = false;
      continue;
    }
    if (!cur) continue;
    const hunk = HUNK.exec(text);
    if (hunk) {
      oldNo = Number(hunk[1]);
      newNo = Number(hunk[2]);
      inHunk = true;
      cur.lines.push({ kind: 'hunk', text, oldNo: null, newNo: null });
      continue;
    }
    if (!inHunk) {
      if (text.startsWith('--- ')) cur.oldPath = stripPrefix(text.slice(4));
      else if (text.startsWith('+++ ')) cur.newPath = stripPrefix(text.slice(4));
      cur.path = cur.newPath || cur.oldPath;
      if (text !== '') cur.header.push(text);
      continue;
    }
    if (text.startsWith('+')) {
      cur.lines.push({ kind: 'add', text: text.slice(1), oldNo: null, newNo: newNo++ });
      cur.added++;
    } else if (text.startsWith('-')) {
      cur.lines.push({ kind: 'del', text: text.slice(1), oldNo: oldNo++, newNo: null });
      cur.removed++;
    } else if (text.startsWith(' ')) {
      cur.lines.push({ kind: 'ctx', text: text.slice(1), oldNo: oldNo++, newNo: newNo++ });
    } else if (text.startsWith('\\')) {
      cur.lines.push({ kind: 'meta', text, oldNo: null, newNo: null });
    }
  }
  return files;
}
